crawl-ref/source/timed-effects.cc

```diff
@@ -1258,3 +1259,104 @@ int speed_to_duration(int speed)
 
     return div_rand_round(100, speed);
 }
+
+static int& _zot_clock_for(branch_type br)
+{
+    static const string CLOCK_KEY_ROOT = "ZOT_CLOCK_FOR_";
+    const string branch_name = branches[br].abbrevname;
+    return you.props[CLOCK_KEY_ROOT + branch_name].get_int();
+}
+
+static int& _zot_clock() {
+    return _zot_clock_for(you.where_are_you);
+}
+
+static bool _zot_clock_active_in(branch_type br)
+{
+    return br != BRANCH_ABYSS && !player_has_orb();
+}
+
+// Is the zot clock running, or is it paused or stopped altogether?
+bool zot_clock_active() {
+    return _zot_clock_active_in(you.where_are_you);
+}
+
+static bool _over_zot_threshold(branch_type br) {
+    return _zot_clock_for(br) >= MAX_ZOT_CLOCK - BEZOTTING_THRESHOLD;
+}
+
+// If the player was in the given branch, would they suffer penalties for
+// nearing the end of the zot clock?
+bool bezotted_in(branch_type br)
+{
+    return _zot_clock_active_in(br) && _over_zot_threshold(br);
+}
+
+// Is the player suffering penalties from nearing the end of the zot clock?
+bool bezotted() {
+    return bezotted_in(you.where_are_you);
+}
+
+// How many times should the player have been drained by Zot?
+int bezotting_level() {
+    if (!bezotted())
+        return 0;
+    const int MAX_ZOTS = 5;
+    const int TURNS_PER_ZOT = (MAX_ZOT_CLOCK - BEZOTTING_THRESHOLD) / MAX_ZOTS;
+    const int over_thresh = _zot_clock() - (MAX_ZOT_CLOCK - BEZOTTING_THRESHOLD);
+    return over_thresh / TURNS_PER_ZOT + 1;
+}
+
+// Decrease the zot clock when the player enters a new level.
+void decr_zot_clock() {
+    if (!zot_clock_active()) {
+        return;
+    }
+    if (bezotted()) {
+        mpr("As you enter the new level, Zot loses track of you.");
+    }
+    int &zot = _zot_clock();
+    zot = max(0, zot - ZOT_CLOCK_PER_FLOOR);
+}
+
+// Odds of the zot clock incrementing every aut, expressed as odds
+// out of 1000 (aka 10x a percent chance).
+static unsigned _zot_clock_odds() {
+    const int base_odds = 100; // 10% per aut, aka on average 1/turn
+    if (have_passive(passive_t::slow_zot)) {
+        // down to 6.7% at full piety, aka once every 1.5 turns. (only movement
+        // (is slowed, not all actions, so we shouldn't give double clock!)
+        return base_odds - div_rand_round(you.piety, 6);
+    }
+    return base_odds;
+}
+
+void incr_zot_clock() {
+    const int clock_incr = binomial(you.time_taken, _zot_clock_odds(), 1000);
+    const int old_lvl = bezotting_level();
+    _zot_clock() += clock_incr;
+    if (!bezotted()) return;
+
+    if (_zot_clock() >= MAX_ZOT_CLOCK)
+    {
+        mpr("Zot has found you!");
+        ouch(INSTANT_DEATH, KILLED_BY_ZOT);
+        return;
+    }
+
+    if (!old_lvl) {
+        mpr("You have lingered too long in familiar places. Zot approaches. Travel to new levels before it's too late!");
```

Comment:
perhaps it would be fair to give at least one warning without consequences before the first drain effect? Maybe at threshold - 100 or something? It might even be worth considering having the status light start early with a different color.

Reply:
I considered that, but it just means that you can scum completely consequence free until you hit the timer. Seems counterproductive.

Reply:
Perhaps it would be even more fair to just display it. Having it a mystery just produces constant anxiety: even if you're still 10,000 turns from being flogged by Son of Food Clock, how would you know? The reasons for even implementing this clock are mysterious to me, even after reading your explanations. Is it about trapfinding? If so, how does it even discourage that playstyle? Trapfinding is more far more wasteful of real time than turns.

In my view, time limits are not intrinsically more fun, and Son of Food Clock is worse than Food Clock on two counts: one, it's entirely implicit, the Food Clock could be monitored simply by keeping track of how many rations you have. two, it even affects undead, for whom previously freedom from the food clock was one of very few selling points. If you merge this, and really I think you shouldn't, it increases the urgency for some kind of buff for non-eating species.

Reply:
Hi! Thanks for the feedback. However, you're posting this as a followup to a specific PR review comment, which is not a good place to post general feedback and background questions. If you want to have this discussion in this PR (I'll let PF decide whether that is appropriate) please repost or followup as a general comment, not in this review thread.

I'll add that clocks are something where there's been an awful lot of discussion both by the community and the devteam, so as usual this PR is not really in a vacuum, and it may be worth looking through some of that before commenting further, as I do think this comment contains some incorrect assumptions about both what this PR is about and what it would do, as well as what the food clock (rip) actually did. The main targets of this are behaviors in the tedious but optimal category that won't resemble a "typical" dcss playstyle at all: extremely excessive luring (i.e. 10s of thousands of turns), certain other kinds of waitscumming (i.e. placing a very outer limit on what kind of xomscumming is possible). One recent community discussion that has a useful summary is this tavern thread from hellmonk: https://crawl.develz.org/tavern/viewtopic.php?f=8&t=26941

Reply:
> I considered that, but it just means that you can scum completely consequence free until you hit the timer. Seems counterproductive.

Unless I'm missing something, can't Hypothetically Optimal Man just track the state of the bank by doing a little arithmetic? In which case, this seems like a thing a spoiled player can know by doing some tedious tracking, and my expectation would be that DCSS just tells me such a thing. I can't see any random element for non-Chei which would prevent it being tracked exactly by HOM.

Reply:
> I can't see any random element for non-Chei which would prevent it being tracked exactly by HOM.

The clock increment currently has some fuzzing to avoid this, see `incr_zot_clock()`. (For clarity's sake, my review comment here was only concerned with how to warn players that they are getting close to a limit.)